src/java/org/apache/cassandra/index/sai/disk/v1/VectorIndexSearcher.java

```diff
@@ -69,30 +76,71 @@ public long indexFileCacheSize()
+    public RangeIterator<Long> searchSSTableRowIds(Expression exp, AbstractBounds<PartitionPosition> keyRange, SSTableQueryContext context, boolean defer, int limit) throws IOException
     {
-        ReorderingPostingList results = searchPosting(exp, limit);
+        ReorderingPostingList results = searchPosting(exp, keyRange, limit);
         return toSSTableRowIdsIterator(results, context);
     }
 
-    private ReorderingPostingList searchPosting(Expression exp, int limit)
+    private ReorderingPostingList searchPosting(Expression exp, AbstractBounds<PartitionPosition> keyRange, int limit)
     {
         if (logger.isTraceEnabled())
             logger.trace(indexContext.logMessage("Searching on expression '{}'..."), exp);
 
         if (exp.getOp() != Expression.Op.ANN)
             throw new IllegalArgumentException(indexContext.logMessage("Unsupported expression during ANN index query: " + exp));
 
+        SparseFixedBitSet bits = bitsetForKeyRange(keyRange);
+
         ByteBuffer buffer = exp.lower.value.raw;
         float[] queryVector = (float[])indexContext.getValidator().getSerializer().deserialize(buffer.duplicate());
-        return graph.search(queryVector, limit, null, Integer.MAX_VALUE);
+        return graph.search(queryVector, limit, bits, Integer.MAX_VALUE);
+    }
+
+    private SparseFixedBitSet bitsetForKeyRange(AbstractBounds<PartitionPosition> keyRange)
+    {
+        // not restricted
+        if (RangeUtil.coversFullRing(keyRange))
+            return null;
+
+        PrimaryKey firstPrimaryKey = keyFactory.createTokenOnly(keyRange.left.getToken());
+        PrimaryKey lastPrimaryKey = keyFactory.createTokenOnly(keyRange.right.getToken());
+
+        // it will return the next row id if given key is not found.
+        long minSSTableRowId = primaryKeyMap.rowIdFromPrimaryKey(firstPrimaryKey);
+        long maxSSTableRowId = primaryKeyMap.lastRowIdFromPrimaryKey(lastPrimaryKey);
+
+        // if it covers entire segment, skip bit set
+        if (minSSTableRowId <= metadata.minSSTableRowId && maxSSTableRowId >= metadata.maxSSTableRowId)
+            return null;
+
+        minSSTableRowId = Math.max(minSSTableRowId, metadata.minSSTableRowId);
+        maxSSTableRowId = Math.min(maxSSTableRowId, metadata.maxSSTableRowId);
+
+        // TODO this is inefficient when range covers most of the segment. Can we change HNSW bit set to filter on row id instead of graph ordinal?
+        SparseFixedBitSet bits = new SparseFixedBitSet(1 + metadata.segmentedRowId(metadata.maxSSTableRowId));
+        for (long sstableRowId = minSSTableRowId; sstableRowId <= maxSSTableRowId; sstableRowId++)
+        {
+            try
+            {
+                int segmentRowId = metadata.segmentedRowId(sstableRowId);
+                // FIXME what if row id doesn't exist in hnsw? e.g. null vector column
```

Comment:
@jbellis is it ok that a given row id doesn't exist in graph? eg. the vector is nullfor the row

Reply:
if it doesn't exist in the graph then it's not part of topK and we should leave it out